Motor Refactor

noaht11 opened the pull request:
**General objectives of this PR**
- Extracting configuration to the top-level
- Making more use of hardware
- Removing duplicated code
- General cleanup and documentation

**Specific changes**
- Deleted LimitSwitch.h and Encoder.h and moved their relevant pieces into Axis
- The step PWM signal is now generated in hardware by the timer `TIOA` output. This allows an easily configurable duty cycle and reduces the requirements for the velocity interrupt
- The velocity interrupt now just updates the counter reset value (i.e. the frequency) of the timer for the step signal
- Debouncing the limit switches in now handled in hardware (a feature of the Due PIO)
- Extracted all the duplicated code between the axes into common inline functions
- Added validation checks to `start_axis` to check all the required conditions are met before starting to move an axis
- The homing routine is now handled by the top-level rather than within Axis to keep the scope of Axis down and eliminate cyclic dependencies. This also simplifies the limit switch ISRs so now they just stop movement and update a state variable
- Deconstructed the `Axis` struct into several smaller structs with well-defined purposes.
- The X and Y instances of the `Axis` struct are now internal to Axis.cpp and external access is accomplished using `AxisId` values
- Configuration of which pins are used by the axes is now controllable from the top-level via the `AxisIO` struct
- The movement information passed to an axis is now overspecified since it includes the holding velocity; however, the axis will stop accelerating once it reaches the holding velocity minimizing the accumulated encoder count error (since we will be guaranteed to be holding at the correct velocity)
- Constrained `Kinematics` to operate in pure math, independent of specific units. Unit conversion (steps to/from counts etc.) are now handled in `Movement`
- Some default movement parameters and encoder count / motor step specifications are now configurable from the top-level via the `mPMTTestStandMotionConfig` struct
- Added periodic dump of axis state information to the debug serial
- Cleaned up top-level state machine handling of axis states
- Added comment documentation to almost all of the Gantry modules

noaht11 committed:
Refactored everything into many structs, inline functions, macros etc.

noaht11 committed:
Finished refactoring, need to fix compiler errors

noaht11 committed:
Compiles and runs, need to debug the timers and interrupts

noaht11 committed:
Fixed timer frequency, still lots of room for optimization

noaht11 committed:
Working using PWM pins

noaht11 committed:
Basic move commands are fully working, no encoder error, fixed get_data offset bug

noaht11 committed:
Fixed limit switches, added hardware debouncing

noaht11 committed:
Added lots of doc comments

noaht11 committed:
Homing works, switched encoder count to signed integer, velocity overshoot problem

noaht11 committed:
Moved back to things in accel interrupt - works need to cleanup

noaht11 committed:
Cleanup, made duty cycle a parameter

noaht11 committed:
Homing routing working!

noaht11 committed:
General cleanup of debug printing

noaht11 committed:
Fixed comment

noaht11 committed:
Extracted motion configs to top-level and now accelerate exactly to hold_vel

noaht11 committed:
Changed ratio back to normal

chuandudx commented on MessageTerminal/MessageTerminal.cxx:
```diff
@@ -99,7 +99,6 @@ void get_status(istringstream& iss)
                 case STATUS_MOVING: puts("Status: MOVING"); break;
                 case STATUS_HOMING: puts("Status: HOMING"); break;
-                case STATUS_FAULT:  puts("Status: FAULT"); break;
                 default:            puts("ERR: Invalid status"); break;
```
What's the reason for removing the fault status?

noaht11 replied:
Oops I removed it then added it back everywhere except here. Will add it back here as well.

noaht11 replied:
Will address this in future PR

chuandudx commented on firmware/lib/Gantry/Axis.cpp:
```diff
@@ -1,311 +1,569 @@
+#define SCLK_FREQ              32768
+
+/** Minimum pulse duration to pass debouncing */
+#define DEBOUNCE_FILTER_MS     50
```
Is this with regards to the encoder? We are waiting 50 milliseconds for what action?

noaht11 replied:
This is for debouncing the limit switches. The limit switch value has to stay steady for 50 ms to be accepted

chuandudx commented on firmware/lib/Gantry/Axis.cpp:
```diff
@@ -1,311 +1,569 @@
+#define REACHED_TARGET(_dir, _cur, _tar)             \
+    (((_dir) == DIR_POSITIVE && (_cur) >= (_tar)) || \
+    ((_dir) == DIR_NEGATIVE && (_cur) <= (_tar)))
+
```
maximum macro-tude

chuandudx commented on firmware/lib/Gantry/Axis.cpp:
```diff
@@ -1,311 +1,569 @@
+    ((_dir) == DIR_NEGATIVE && (_cur) <= (_tar)))
+
+/** Maximum allowed velocity for an axis */
+#define VEL_MAX                25000
```
Can we add its units in comment or in the variable name for clarity. In this case, it looks like counts?

noaht11 replied:
It's actually steps / second here, so evidently good point. I actually realize I need to move this up a couple levels, so will do that and add units in a comment

chuandudx commented on firmware/lib/Gantry/Axis.h:
```diff
@@ -1,69 +1,108 @@
+typedef enum {
+    AXIS_OK,                           //!< Axis movement started OK
+    AXIS_ERR_DIR_MISMATCH,             //!< Direction did not match count value signs
+    AXIS_ERR_ZERO_DIST,                //!< Total specified distance was zero
```
Does specifying zero distance qualify as an error? Or should we just not move that axis.

noaht11 replied:
I guess there's nothing wrong with accepting a zero distance movement. I can switch it to not be an error

chuandudx commented on firmware/lib/Gantry/Kinematics.h:
```diff
@@ -4,12 +4,12 @@
-    uint32_t counts_accel;
-    uint32_t counts_hold;
-    uint32_t counts_decel;
+    uint32_t dist_accel;
```
Similar comment - can we clarify the unit of `dist` (counts or mm)?

noaht11 replied:
Ahh, it's described in the comments on `generate_vel_profile` in `Kinematics.cpp`. The units are actually intentionally not specified because it's up to the user. I made it so this `Kinematics` module deals in pure math where it'll use whatever units are used for parameters that get passed to functions.
I can add some comments here to clarify that.

chuandudx commented on firmware/lib/Gantry/Movement.cpp:
```diff
@@ -1,54 +1,64 @@
+ *                      encoder counts and motor steps
+ * 
+ * @return true if the axis successfully started executing the motion, otherwise false
+ */
```
nice description!

chuandudx commented on firmware/lib/Gantry/Movement.cpp:
```diff
@@ -1,54 +1,64 @@
+
+bool move_axis_home_b(AxisId axis_id, MotionShape shape, MotionUnits units)
+{
+    return move_axis_rel(axis_id, DIR_POSITIVE, INT32_MAX, shape, units);
```
This is always specifying a max possible number of count to move, until the limit switch is reached?

noaht11 replied:
Yeah, the idea is try to move forever, the limit switch will automatically stop you once you've reached the home.

chuandudx commented on firmware/lib/Gantry/Timer.cpp:
```diff
@@ -1,56 +1,151 @@
+ * | TIMER_CLOCK4 | MCK/128    |
+ * | TIMER_CLOCK5 | SLCK       |
+ */
+
```
Nice addition!

chuandudx commented on firmware/lib/Gantry/Timer.h:
```diff
@@ -1,14 +1,24 @@
 
-void start_timer_accel(Tc *tc, uint32_t channel, IRQn_Type irq, uint32_t accel);
+// PWM Timers
+void configure_pwm_timer(Tc *tc, uint32_t channel, IRQn_Type irq, Pio *pio_bank, EPioType periph, uint32_t pin_mask);
```
Just realizing that all the functions use underscore convention as opposed to camelCase?

noaht11 replied:
Most of the C code I've ever dealt with uses `snake_case`.
Idk what the C++ convention is, but all of the Axis code is actually pure C so I think we should stick with snake_case here.

chuandudx commented on firmware/src/mPMTTestStand.cpp:
```diff
@@ -111,34 +112,108 @@ void mPMTTestStand::handle_get_data()
+            }
+            else if (this->x_state->ls_home_pressed || this->x_state->ls_far_pressed ||
+                     this->y_state->ls_home_pressed || this->x_state->ls_far_pressed) {
+                this->status = STATUS_FAULT;
```
Did we remove the fault status earlier on?

noaht11 replied:
It was removed, then re-added, but I forgot it in that other spot, see the comment above.

noaht11 commented on firmware/lib/Gantry/Kinematics.cpp:
```diff
@@ -1,38 +1,83 @@
+// {
+//     uint32_t elapsed_steps = elapsed_counts * MOTOR_STEPS_PER_REV / ENCODER_COUNTS_PER_REV;
+//     return (uint32_t)sqrt((2 * accel * elapsed_steps) + (v_min * v_min));
+// }
```
Need to delete this

noaht11 commented on firmware/lib/Gantry/Kinematics.h:
```diff
@@ -4,12 +4,12 @@
 
-uint32_t calc_counts_accel(uint32_t accel, uint32_t v_min, uint32_t v_max);
-bool generate_vel_profile(uint32_t accel, uint32_t v_min, uint32_t v_max, uint32_t counts_total, VelProfile *profile_out);
+// uint32_t calc_expected_velocity(uint32_t accel, uint32_t v_min, uint32_t elapsed_counts);
```
Need to delete

chuandudx commented:
As a general question for the limit switches, are they already initialized with `INPUT_PULLUP`. If so, I will exclude a hardware resistor pull-up from schematic to simplify the soldering job required.

alexproskurin commented on firmware/lib/Gantry/Axis.cpp:
```diff
@@ -1,311 +1,569 @@
+void isr_ls_far_x();
+
+static Axis axis_x = {
+    .io = {},
```
Didn't know you could do this dot thing, pretty sweet!

noaht11 replied:
Yup, great way to initialize structs!

noaht11 commented:
> As a general question for the limit switches, are they already initialized with `INPUT_PULLUP`. If so, I will exclude a hardware resistor pull-up from schematic to simplify the soldering job required.

@chuandudx Yeah they're internally pulled up so do not need an external pull-up

chuandudx commented:
consider removing pin assignments `ENCODER_PIN_B_X` and `ENCODER_PIN_B_Y` since they aren't used.

alexproskurin commented on firmware/lib/Gantry/Axis.cpp:
```diff
@@ -1,311 +1,569 @@
-    delay(1000);
-    attachInterrupt(digitalPinToInterrupt(axis_config->ls_home), axis_config->isr_limit_switch, CHANGE);
-    delay(1000);
-    attachInterrupt(digitalPinToInterrupt(axis_config->ls_far), axis_config->isr_limit_switch, CHANGE);
```
I don't think this one should be set up to trigger on `CHANGE`. I think we should only trigger it when it is first pressed, so `FALLING`. I think the current set up breaks the homing routine in a case when we have `ls_far` `PRESSED` and we start homing.

noaht11 replied:
Sounds good

alexproskurin commented on firmware/lib/Gantry/Axis.cpp:
```diff
@@ -1,311 +1,569 @@
+/*****************************************************************************/
+
+/** TC IRQ number for x-axis velocity timer */
+#define IRQ_X_AXIS_VELOCITY    3
```
Kinda nitpicking here, I think a better naming convention is: `IRQ_AXIS_VELOCITY_X` for X and `IRQ_AXIS_VELOCITY_Y` for Y, that is having a differentiating identifier at the end of the variable name.

noaht11 replied:
I like it (as in your suggestion)

noaht11 replied:
Actually I changed my mind. I picture the hierarchy as:

- Gantry
  - X Axis
    - Velocity Interrupt
    - Accel Interrupt
    - Other Interrupts...
  - Y Axis
    - Velocity Interrupt
    - Accel Interrupt
    - Other Interrupts...

So I'm narrowing down in order of hierarchy: first identify the axis, then which interrupt

Not opposed to switching to `IRQ_AXIS_X_VELOCITY` and `IRQ_AXIS_Y_VELOCITY` etc.

alexproskurin replied:
I could kinda see it go either way haha so i am good with either

noaht11 replied:
Okay I'll go with my current one then to save the extra commit

alexproskurin commented on firmware/lib/Gantry/Axis.cpp:
```diff
@@ -1,311 +1,569 @@
+    setup_debouncing(axis->io.pin_ls_far, DEBOUNCE_FILTER_MS);
+    attachInterrupt(digitalPinToInterrupt(axis->io.pin_ls_far),
+                    axis->interrupts.isr_ls_far,
+                    CHANGE);
```
I think i commented on already removed code for this, but still the same thing - switch `ls_far` from `CHANGE` to `FALLING`

noaht11 replied:
Good idea

noaht11 replied:
Although like you mentioned in the other comment about switching the wiring, whether this is `FALLING` or `RISING` should be defined in the same place as whether pressed is `LOW` or `HIGH`

alexproskurin replied:
Can we not define RISING or FALLING in terms of HIGH or LOW of the unpressed switch? I think internally RISING/FALLING maps to an integer.

noaht11 replied:
I don't wanna depend on the internal definitions of RISING/FALLING (everything should still work if those definitions change).
Anyway now, the choice of RISING/FALLING is derived from the choice of LOW/HIGH for the pressed state

alexproskurin commented on firmware/lib/Gantry/Axis.cpp:
```diff
@@ -1,311 +1,569 @@
+    // Configure state
+    axis->state.moving = true;
+    axis->state.velocity = motion->vel_start;
+    axis->state.next_velocity = axis->state.velocity;
```
What is the next velocity for?

noaht11 replied:
Next velocity is so that if the velocity hasn't changed (e.g. during the holding segment or if we're accelerating slowly compare to our current velocity) we can notice that and not bother resetting the timer in the velocity ISR

alexproskurin commented on firmware/lib/Gantry/Axis.cpp:
```diff
@@ -1,311 +1,569 @@
+        VEL_DUTY_CYCLE);
+
+    // Start acceleration timer interrupt
+    reset_timer_interrupt(
```
Do we only use this for acceleration? Might be more informative to just have it be named `reset_acceleration_timer_interrupt`?

noaht11 replied:
There's nothing in the function itself that is inherent to acceleration. It starts a timer interrupt at some frequency, could be used for anything

noaht11 replied:
There's nothing in the function itself that is inherent to acceleration. It starts a timer interrupt at some frequency, could be used for anything

alexproskurin commented on firmware/lib/Gantry/Axis.cpp:
```diff
@@ -1,311 +1,569 @@
-    last_interrupt_time = interrupt_time;
+    stop_axis(axis);
+    // Limit switch reads LOW when pressed
+    axis->state.ls_home_pressed = (digitalRead(axis->io.pin_ls_home) == LOW);
```
Instead of comparing to `LOW`, would it be easier to read if it was compared to limit switch status enum `PRESSED`. That way if the limit switch wiring ever changes you could just swap the order of limit switch status assignment in enum? Orr is this not very C like?

noaht11 replied:
I made it so that the definition of the "pressed" state is part of the AxisIO, so we can configure it from there

alexproskurin commented on firmware/lib/Gantry/Axis.cpp:
```diff
@@ -1,311 +1,569 @@
+{
+    stop_axis(axis);
+    // Limit switch reads LOW when pressed
+    axis->state.ls_far_pressed = (digitalRead(axis->io.pin_ls_far) == LOW);
```
Same as above

alexproskurin commented on firmware/lib/Gantry/Axis.cpp:
```diff
@@ -1,311 +1,569 @@
+    if (!axis->state.moving) return;
+
+    switch (axis->state.velocity_segment) {
+        case VEL_SEG_ACCELERATE:
```
I think i was reading somewhere that if you have less than 5 switch cases, simple if statements could be faster

noaht11 replied:
switch statements will definitely be faster for 5+ items because they will get compiled into a jump table. For less than 5 items there is negligible difference, so I went with switch cause it seemed more legible

alexproskurin commented on firmware/lib/Gantry/Axis.cpp:
```diff
@@ -1,311 +1,569 @@
+            break;
         }
     }
 }
```
Not really sure if we can achieve the same accuracy without checking encoder counts and just relying on specified velocity. Just seems a little dangerous not to have a secondary check, but i might be missing something that got changed elsewhere

noaht11 replied:
You mean for the accelerate case? Both hold and decelerate check the encoder counts.
At the end of accelerate, I still have your error adjustment so if we reached holding velocity early, we'll hold for longer or if we reached it late, we'll hold for less time, so the accuracy should be the exact same as before, unless I'm missing something?

What this improves is we will never be holding at the wrong velocity. The final exit condition (i.e. the only way to reach stop_axis) is still to have reached to correct final encoder count.

noaht11 replied:
The potential issue is if the hold_vel passed to start_axis doesn't match the counts. I think I'm actually going to eliminate the Movement files and put all that calculation logic into the Axis so it's not dependent on correct values being passed to it

alexproskurin replied:
OK, I think we are good then

alexproskurin commented on firmware/lib/Gantry/Axis.cpp:
```diff
@@ -1,311 +1,569 @@
+ * 
+ * @brief Specifies all of the internal interrupts and ISRs used by the axis
+ */
+typedef struct {
```
How come are these defined in .cpp and not in .h?

noaht11 replied:
They are not used by (and should not be visible to) anyone outside of this .cpp file

noaht11 committed:
Addressed some PR comments

noaht11 committed:
Removed Movement.cpp as a middle man

- MotionSpec's are now sent directly to an axis which handles unit
conversion and generating a profile since it depended on those profiles
being valid anyway.
- Also simplifies error handling since there is only one level at which
errors can be generated (in the Axis)
- Also makes the gantry mech config a one time thing when the axis is
setup

chuandudx commented on firmware/src/main.cpp:
```diff
@@ -1,35 +1,69 @@
+    },
+    .gantry = {
+        .axis_mech = {
+            .counts_per_rev     = 800,    // For Igus gantry encoders
```
IGUS encoders are 500 counts/rev

chuandudx commented on firmware/src/main.cpp:
```diff
@@ -1,35 +1,69 @@
+        .serial_comm_baud_rate  = 115200,
+        // Thermistor Pins
+        .pin_therm_amb          = A0,
+        .pin_therm_motor1       = A1,
```
Can we make `pin_therm_motor1` into `pin_therm_motor_axis_x` and  `pin_therm_motor2` into `pin_therm_motor_axis_y` . I think I was the one to call them motor1 and 2 but it's unclear what they refer to looking back.

noaht11 replied:
How about just `pin_therm_motor_x` and `pin_therm_motor_y` (no `axis` to keep it shorter)

chuandudx replied:
sounds good to me!

noaht11 commented:
> consider removing pin assignments `ENCODER_PIN_B_X` and `ENCODER_PIN_B_Y` since they aren't used.

@chuandudx I'm gonna leave them for now, we can purge them at the very end if we want

noaht11 committed:
Addressed PR comments

Made limit switch active state configurable via AxisIO
Renamed motor thermistor pins
Fixed encoder counts per revolution
